dlt/destinations/impl/clickhouse/sql_client.py

```diff
@@ -0,0 +1,174 @@
+from contextlib import contextmanager
+from typing import (
+    Iterator,
+    AnyStr,
+    Any,
+    Optional,
+    Sequence,
+    ClassVar,
+    Union,
+)
+
+import clickhouse_driver  # type: ignore[import-untyped]
+import clickhouse_driver.errors  # type: ignore[import-untyped]
+from clickhouse_driver.dbapi import Connection  # type: ignore[import-untyped]
+from clickhouse_driver.dbapi.extras import DictCursor  # type: ignore[import-untyped]
+
+from dlt.common.destination import DestinationCapabilitiesContext
+from dlt.common.runtime import logger
+from dlt.destinations.exceptions import (
+    DatabaseUndefinedRelation,
+    DatabaseTransientException,
+    DatabaseTerminalException,
+)
+from dlt.destinations.impl.clickhouse import capabilities
+from dlt.destinations.impl.clickhouse.configuration import ClickhouseCredentials
+from dlt.destinations.sql_client import (
+    DBApiCursorImpl,
+    SqlClientBase,
+    raise_database_error,
+    raise_open_connection_error,
+)
+from dlt.destinations.typing import DBTransaction, DBApi
+
+
+TRANSACTIONS_UNSUPPORTED_WARNING_MESSAGE = (
+    "Clickhouse does not support transactions! Each statement is auto-committed separately."
+)
+
+
+class ClickhouseDBApiCursorImpl(DBApiCursorImpl):
+    native_cursor: DictCursor
+
+
+class ClickhouseSqlClient(
+    SqlClientBase[clickhouse_driver.dbapi.connection.Connection], DBTransaction
+):
+    dbapi: ClassVar[DBApi] = clickhouse_driver.dbapi.connection.Connection
+    capabilities: ClassVar[DestinationCapabilitiesContext] = capabilities()
+
+    def __init__(self, dataset_name: str, credentials: ClickhouseCredentials) -> None:
+        super().__init__(credentials.database, dataset_name)
+        self._conn: clickhouse_driver.dbapi.connection = None
+        self.credentials = credentials
+        self.database_name = credentials.database
+
+    def open_connection(self) -> clickhouse_driver.dbapi.connection.Connection:
+        self._conn = clickhouse_driver.connect(dsn=self.credentials.to_native_representation())
+        return self._conn
+
+    @raise_open_connection_error
+    def close_connection(self) -> None:
+        if self._conn:
+            self._conn.close()
+            self._conn = None
+
+    @contextmanager
+    @raise_database_error
+    def begin_transaction(self) -> Iterator[DBTransaction]:
```

Comment:
I think we may at some point want to a capabiliy wether a destination supports transactions and then deal with this automatically, but maybe not now.

Reply:
Good idea. Please advise if you see any way I can make this PR deal better with CH's lack of transactions, if so.

(As a side note, [Experimental support](https://clickhouse.com/docs/en/guides/developer/transactional#enable-experimental-transaction-suppor) is available for single node deployments, mergetree family)